File: 2024/day03/main.py
#!/usr/bin/env python3

import sys
import re

def part1():
    with open(sys.argv[1], 'r') as file:
        acc = 0
        for line in file:
            matches = re.findall(r"mul\((\d{1,3}),(\d{1,3})\)", line)
            for m in matches:
                acc += int(m[0]) * int(m[1])
        print("Part 1:", acc)

def part2():
    with open(sys.argv[1], 'r') as file:
        acc = 0
        skip = False
        for line in file:
            matches = re.findall(r"(mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\))", line)
            for m in matches:
                if m[0] == 'do()':
                    skip = False
                elif m[0] == 'don\'t()':
                    skip = True
                else:
                    if skip == False:
                        acc += int(m[1]) * int(m[2])
        print("Part 2:", acc)

part1()
part2()

File: 2024/day03/main.pl
#!/usr/bin/env perl

use warnings;
use strict;

sub part1 {
    my @lines = @_;
    my $sum = 0;
    foreach (@lines) {
        while (/mul\((\d{1,3}),(\d{1,3})\)/gc) {
            $sum += $1 * $2;
        }
    }
    print "Part 1: $sum\n";
}

sub part2 {
    my @lines = @_;
    my $sum = 0;
    my $skip = 1;
    foreach (@lines) {
        while (/(mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\))/gc) {
            my ($opa, $opb, $opc) = ($1, $2, $3);
            if ($opa =~ /^mul/) {
                $sum += $opb * $opc * $skip;
            } elsif ($opa =~ /^don/) {
                $skip = 0;
            } elsif ($opa =~ /^do/) {
                $skip = 1;
            }
        }
    }
    print "Part 2: $sum\n";
}

my @input = <>;
&part1(@input);
&part2(@input);